src/core/auth/qgsauthcertutils.cpp

```diff
@@ -191,37 +189,55 @@ QSslKey QgsAuthCertUtils::keyFromFile( const QString &keypath,
                                        const QString &keypass,
                                        QString *algtype )
 {
-  bool pem = keypath.endsWith( QLatin1String( ".pem" ), Qt::CaseInsensitive );
-  QByteArray keydata( QgsAuthCertUtils::fileData( keypath, pem ) );
+  // The approach here is to try all possible encodings and algorithms
+  QByteArray keydata( QgsAuthCertUtils::fileData( keypath ) );
 
   QSslKey clientkey;
   clientkey = QSslKey( keydata,
                        QSsl::Rsa,
-                       pem ? QSsl::Pem : QSsl::Der,
+                       QSsl::Pem,
                        QSsl::PrivateKey,
                        !keypass.isEmpty() ? keypass.toUtf8() : QByteArray() );
-  if ( clientkey.isNull() )
-  {
-    // try DSA algorithm, since Qt can't seem to determine it otherwise
-    clientkey = QSslKey( keydata,
-                         QSsl::Dsa,
-                         pem ? QSsl::Pem : QSsl::Der,
-                         QSsl::PrivateKey,
-                         !keypass.isEmpty() ? keypass.toUtf8() : QByteArray() );
-    if ( clientkey.isNull() )
-    {
-      return QSslKey();
-    }
+  if ( ! clientkey.isNull() )
+  {
+    if ( algtype )
+      *algtype = QStringLiteral( "rsa" );
+    return clientkey;
+  }
+  clientkey = QSslKey( keydata,
+                       QSsl::Dsa,
+                       QSsl::Pem,
+                       QSsl::PrivateKey,
+                       !keypass.isEmpty() ? keypass.toUtf8() : QByteArray() );
+  if ( ! clientkey.isNull() )
+  {
     if ( algtype )
       *algtype = QStringLiteral( "dsa" );
+    return clientkey;
   }
-  else
+  clientkey = QSslKey( keydata,
+                       QSsl::Rsa,
+                       QSsl::Der,
+                       QSsl::PrivateKey,
+                       !keypass.isEmpty() ? keypass.toUtf8() : QByteArray() );
+  if ( ! clientkey.isNull() )
   {
     if ( algtype )
       *algtype = QStringLiteral( "rsa" );
+    return clientkey;
   }
-
-  return clientkey;
+  clientkey = QSslKey( keydata,
+                       QSsl::Dsa,
+                       QSsl::Der,
+                       QSsl::PrivateKey,
+                       !keypass.isEmpty() ? keypass.toUtf8() : QByteArray() );
+  if ( ! clientkey.isNull() )
+  {
+    if ( algtype )
+      *algtype = QStringLiteral( "dsa" );
+    return clientkey;
+  }
+  return QSslKey();
 }
```

Comment:
We need to also test for `QSsl::Ec (The Elliptic Curve algorithm)`, though not sure how to test for `QSsl::Opaque` -- or not test for `QSsl::Opaque` and just return the empty QSslKey.

(see [QSsl::KeyAlgorithm](http://doc.qt.io/qt-5/qssl.html#KeyAlgorithm-enum))

Reply:
Question is: if the type is Qssl::Opaque would it be treated differently than an empty/null key within QGIS?

Reply:
I meant leave it as unsupported. However, best to leave how an opaque key is handled to Qt, since that is the point of that alg type, e.g. PKSC#11 smart card support, etc.